<a href="https://colab.research.google.com/github/kuheli31/Al-Medical-Assistant/blob/main/Symptoms_Disease.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**This  notebook focuses on predicting disease based on symptoms given**

In [51]:
#Importing Dependencies

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sn
from sklearn import metrics
from sklearn.model_selection import train_test_split

In [3]:
#Loading the dataset
df = pd.read_csv('/content/Datasets/dataset.csv')
df

Disease          Symptom_1              Symptom_2  \
0     Fungal infection            itching              skin_rash   
1     Fungal infection          skin_rash   nodal_skin_eruptions   
2     Fungal infection            itching   nodal_skin_eruptions   
3     Fungal infection            itching              skin_rash   
4     Fungal infection            itching              skin_rash   
...                ...                ...                    ...   
4920  Fungal infection          skin rash                itching   
4921  Fungal infection           itching               skin rash   
4922          Impetigo          skin rash             high fever   
4923          Impetigo          skin rash             high fever   
4924          Impetigo  yellow crust ooze   red sore around nose   

                  Symptom_3             Symptom_4          Symptom_5  \
0      nodal_skin_eruptions   dischromic _patches                NaN   
1       dischromic _patches                   NaN                NaN   
2       dischromic _patches                   NaN                NaN   
3       dischromic _patches                   NaN                NaN   
4      nodal_skin_eruptions                   NaN                NaN   
...                     ...                   ...                ...   
4920   nodal skin eruptions                     0                  0   
4921   nodal skin eruptions                     0                  0   
4922                blister  red sore around nose  yellow crust ooze   
4923                blister  red sore around nose  yellow crust ooze   
4924                blister            high fever          skin rash   

     Symptom_6 Symptom_7 Symptom_8 Symptom_9 Symptom_10 Symptom_11 Symptom_12  \
0          NaN       NaN       NaN       NaN        NaN        NaN        NaN   
1          NaN       NaN       NaN       NaN        NaN        NaN        NaN   
2          NaN       NaN       NaN       NaN        NaN        NaN        NaN   
3          NaN       NaN       NaN       NaN        NaN        NaN        NaN   
4          NaN       NaN       NaN       NaN        NaN        NaN        NaN   
...        ...       ...       ...       ...        ...        ...        ...   
4920         0         0         0         0          0          0          0   
4921         0         0         0         0          0          0          0   
4922         0         0         0         0          0          0          0   
4923         0         0         0         0          0          0          0   
4924         0         0         0         0          0          0          0   

     Symptom_13 Symptom_14 Symptom_15 Symptom_16 Symptom_17  
0           NaN        NaN        NaN        NaN        NaN  
1           NaN        NaN        NaN        NaN        NaN  
2           NaN        NaN        NaN        NaN        NaN  
3           NaN        NaN        NaN        NaN        NaN  
4           NaN        NaN        NaN        NaN        NaN  
...         ...        ...        ...        ...        ...  
4920          0          0          0          0          0  
4921          0          0          0          0          0  
4922          0          0          0          0          0  
4923          0          0          0          0          0  
4924          0          0          0          0          0  

[4925 rows x 18 columns]

In [4]:
#Checking how many types of disease presents here?
df["Disease"].value_counts()

Disease
Impetigo                                   123
Fungal infection                           122
Allergy                                    120
Chronic cholestasis                        120
GERD                                       120
Peptic ulcer diseae                        120
AIDS                                       120
Diabetes                                   120
Drug Reaction                              120
Gastroenteritis                            120
Bronchial Asthma                           120
Migraine                                   120
Hypertension                               120
Paralysis (brain hemorrhage)               120
Jaundice                                   120
Malaria                                    120
Cervical spondylosis                       120
Dengue                                     120
Typhoid                                    120
hepatitis A                                120
Hepatitis B                                120
Hepatitis C                                120
Hepatitis D                                120
Hepatitis E                                120
Chicken pox                                120
Alcoholic hepatitis                        120
Tuberculosis                               120
Pneumonia                                  120
Common Cold                                120
Heart attack                               120
Varicose veins                             120
Hypothyroidism                             120
Dimorphic hemmorhoids(piles)               120
Hyperthyroidism                            120
Hypoglycemia                               120
Arthritis                                  120
Osteoarthristis                            120
(vertigo) Paroymsal  Positional Vertigo    120
Acne                                       120
Urinary tract infection                    120
Psoriasis                                  120
Name: count, dtype: int64

In [5]:
df.describe()

Disease  Symptom_1  Symptom_2 Symptom_3    Symptom_4  Symptom_5  \
count       4925       4925       4925      4925         4577       3719   
unique        41         37         52        56           53         41   
top     Impetigo   vomiting   vomiting   fatigue   high_fever   headache   
freq         123        822        870       726          378        348   

       Symptom_6        Symptom_7        Symptom_8           Symptom_9  \
count       2939             2273             1949                1697   
unique        33               27               22                  23   
top       nausea   abdominal_pain   abdominal_pain   yellowing_of_eyes   
freq         390              264              276                 228   

                Symptom_10     Symptom_11 Symptom_12    Symptom_13  \
count                 1517           1199        749           509   
unique                  22             19         12             9   
top      yellowing_of_eyes   irritability    malaise   muscle_pain   
freq                   198            120        126            72   

         Symptom_14   Symptom_15        Symptom_16    Symptom_17  
count           311          245               197            77  
unique            5            4                 4             2  
top      chest_pain   chest_pain   blood_in_sputum   muscle_pain  
freq             96          144                72            72

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4925 entries, 0 to 4924
Data columns (total 18 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Disease     4925 non-null   object
 1   Symptom_1   4925 non-null   object
 2   Symptom_2   4925 non-null   object
 3   Symptom_3   4925 non-null   object
 4   Symptom_4   4577 non-null   object
 5   Symptom_5   3719 non-null   object
 6   Symptom_6   2939 non-null   object
 7   Symptom_7   2273 non-null   object
 8   Symptom_8   1949 non-null   object
 9   Symptom_9   1697 non-null   object
 10  Symptom_10  1517 non-null   object
 11  Symptom_11  1199 non-null   object
 12  Symptom_12  749 non-null    object
 13  Symptom_13  509 non-null    object
 14  Symptom_14  311 non-null    object
 15  Symptom_15  245 non-null    object
 16  Symptom_16  197 non-null    object
 17  Symptom_17  77 non-null     object
dtypes: object(18)
memory usage: 692.7+ KB


In [8]:
#Checking how many categories are there in each and every column
for col in df.columns:
  print(col, ':', len(df[col].unique()) ,'labels')

Disease : 41 labels
Symptom_1 : 37 labels
Symptom_2 : 52 labels
Symptom_3 : 56 labels
Symptom_4 : 54 labels
Symptom_5 : 42 labels
Symptom_6 : 34 labels
Symptom_7 : 28 labels
Symptom_8 : 23 labels
Symptom_9 : 24 labels
Symptom_10 : 23 labels
Symptom_11 : 20 labels
Symptom_12 : 13 labels
Symptom_13 : 10 labels
Symptom_14 : 6 labels
Symptom_15 : 5 labels
Symptom_16 : 5 labels
Symptom_17 : 3 labels


In [9]:
df

Disease          Symptom_1              Symptom_2  \
0     Fungal infection            itching              skin_rash   
1     Fungal infection          skin_rash   nodal_skin_eruptions   
2     Fungal infection            itching   nodal_skin_eruptions   
3     Fungal infection            itching              skin_rash   
4     Fungal infection            itching              skin_rash   
...                ...                ...                    ...   
4920  Fungal infection          skin rash                itching   
4921  Fungal infection           itching               skin rash   
4922          Impetigo          skin rash             high fever   
4923          Impetigo          skin rash             high fever   
4924          Impetigo  yellow crust ooze   red sore around nose   

                  Symptom_3             Symptom_4          Symptom_5  \
0      nodal_skin_eruptions   dischromic _patches                NaN   
1       dischromic _patches                   NaN                NaN   
2       dischromic _patches                   NaN                NaN   
3       dischromic _patches                   NaN                NaN   
4      nodal_skin_eruptions                   NaN                NaN   
...                     ...                   ...                ...   
4920   nodal skin eruptions                     0                  0   
4921   nodal skin eruptions                     0                  0   
4922                blister  red sore around nose  yellow crust ooze   
4923                blister  red sore around nose  yellow crust ooze   
4924                blister            high fever          skin rash   

     Symptom_6 Symptom_7 Symptom_8 Symptom_9 Symptom_10 Symptom_11 Symptom_12  \
0          NaN       NaN       NaN       NaN        NaN        NaN        NaN   
1          NaN       NaN       NaN       NaN        NaN        NaN        NaN   
2          NaN       NaN       NaN       NaN        NaN        NaN        NaN   
3          NaN       NaN       NaN       NaN        NaN        NaN        NaN   
4          NaN       NaN       NaN       NaN        NaN        NaN        NaN   
...        ...       ...       ...       ...        ...        ...        ...   
4920         0         0         0         0          0          0          0   
4921         0         0         0         0          0          0          0   
4922         0         0         0         0          0          0          0   
4923         0         0         0         0          0          0          0   
4924         0         0         0         0          0          0          0   

     Symptom_13 Symptom_14 Symptom_15 Symptom_16 Symptom_17  
0           NaN        NaN        NaN        NaN        NaN  
1           NaN        NaN        NaN        NaN        NaN  
2           NaN        NaN        NaN        NaN        NaN  
3           NaN        NaN        NaN        NaN        NaN  
4           NaN        NaN        NaN        NaN        NaN  
...         ...        ...        ...        ...        ...  
4920          0          0          0          0          0  
4921          0          0          0          0          0  
4922          0          0          0          0          0  
4923          0          0          0          0          0  
4924          0          0          0          0          0  

[4925 rows x 18 columns]

In [11]:
#Finding most frequent values
df.value_counts().sort_values(ascending=False)

Disease           Symptom_1             Symptom_2             Symptom_3             Symptom_4             Symptom_5          Symptom_6  Symptom_7             Symptom_8  Symptom_9  Symptom_10          Symptom_11        Symptom_12       Symptom_13   Symptom_14   Symptom_15   Symptom_16      Symptom_17  
Common Cold        continuous_sneezing   chills                fatigue               cough                 high_fever         headache   swelled_lymph_nodes   malaise    phlegm     throat_irritation   redness_of_eyes   sinus_pressure   runny_nose   congestion   chest_pain   loss_of_smell   muscle_pain    72
Impetigo          skin rash             high fever            blister               red sore around nose  yellow crust ooze  0          0                     0          0          0                   0                 0                0            0            0            0               0                2
Fungal infection  itching               skin rash             nodal skin eruptions  0                     0                  0          0                     0          0          0                   0                 0                0            0            0            0               0                1
                  skin rash             itching               nodal skin eruptions  0                     0                  0          0                     0          0          0                   0                 0                0            0            0            0               0                1
Impetigo          yellow crust ooze     red sore around nose  blister               high fever            skin rash          0          0                     0          0          0                   0                 0                0            0            0            0               0                1
Name: count, dtype: int64

In [17]:
# Converting Categorical datas into numerical one
for col in df.columns:
    df[col] = df[col].str.replace('_',' ')

# Removing the white space in each cell of Dataframe
cols = df.columns
data = df[cols].values.flatten()

s = pd.Series(data)
s = s.str.strip()
s = s.values.reshape(df.shape)

dataset = pd.DataFrame(s,columns = cols)
dataset.head()

Disease  Symptom_1             Symptom_2             Symptom_3  \
0  Fungal infection    itching             skin rash  nodal skin eruptions   
1  Fungal infection  skin rash  nodal skin eruptions   dischromic  patches   
2  Fungal infection    itching  nodal skin eruptions   dischromic  patches   
3  Fungal infection    itching             skin rash   dischromic  patches   
4  Fungal infection    itching             skin rash  nodal skin eruptions   

             Symptom_4 Symptom_5 Symptom_6 Symptom_7 Symptom_8 Symptom_9  \
0  dischromic  patches       NaN       NaN       NaN       NaN       NaN   
1                  NaN       NaN       NaN       NaN       NaN       NaN   
2                  NaN       NaN       NaN       NaN       NaN       NaN   
3                  NaN       NaN       NaN       NaN       NaN       NaN   
4                  NaN       NaN       NaN       NaN       NaN       NaN   

  Symptom_10 Symptom_11 Symptom_12 Symptom_13 Symptom_14 Symptom_15  \
0        NaN        NaN        NaN        NaN        NaN        NaN   
1        NaN        NaN        NaN        NaN        NaN        NaN   
2        NaN        NaN        NaN        NaN        NaN        NaN   
3        NaN        NaN        NaN        NaN        NaN        NaN   
4        NaN        NaN        NaN        NaN        NaN        NaN   

  Symptom_16 Symptom_17  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [20]:
#Filling missing value with 0
dataset.fillna(0, inplace = True)
dataset.head()

Disease  Symptom_1             Symptom_2             Symptom_3  \
0  Fungal infection    itching             skin rash  nodal skin eruptions   
1  Fungal infection  skin rash  nodal skin eruptions   dischromic  patches   
2  Fungal infection    itching  nodal skin eruptions   dischromic  patches   
3  Fungal infection    itching             skin rash   dischromic  patches   
4  Fungal infection    itching             skin rash  nodal skin eruptions   

             Symptom_4 Symptom_5 Symptom_6 Symptom_7 Symptom_8 Symptom_9  \
0  dischromic  patches         0         0         0         0         0   
1                    0         0         0         0         0         0   
2                    0         0         0         0         0         0   
3                    0         0         0         0         0         0   
4                    0         0         0         0         0         0   

  Symptom_10 Symptom_11 Symptom_12 Symptom_13 Symptom_14 Symptom_15  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

  Symptom_16 Symptom_17  
0          0          0  
1          0          0  
2          0          0  
3          0          0  
4          0          0

**Symptoms severity(Symptom severity refers to how intense or serious a symptom is. It helps measure how much a symptom affects a person's health, comfort, or ability to function.)**

In [22]:
severity = pd.read_csv('/content/Datasets/Symptom-severity.csv')
severity

Symptom  weight
0                 itching       1
1               skin_rash       3
2    nodal_skin_eruptions       4
3     continuous_sneezing       4
4               shivering       5
..                    ...     ...
128    inflammatory_nails       2
129               blister       4
130  red_sore_around_nose       2
131     yellow_crust_ooze       3
132             prognosis       5

[133 rows x 2 columns]

In [23]:
severity.value_counts()

Symptom                weight
abdominal_pain         4         1
abnormal_menstruation  6         1
acidity                3         1
acute_liver_failure    6         1
altered_sensorium      2         1
                                ..
weight_loss            3         1
yellow_crust_ooze      3         1
yellow_urine           4         1
yellowing_of_eyes      4         1
yellowish_skin         3         1
Name: count, Length: 133, dtype: int64

In [24]:
# remove the '_' from the dataset
severity['Symptom'] = severity['Symptom'].str.replace('_',' ')
severity

Symptom  weight
0                 itching       1
1               skin rash       3
2    nodal skin eruptions       4
3     continuous sneezing       4
4               shivering       5
..                    ...     ...
128    inflammatory nails       2
129               blister       4
130  red sore around nose       2
131     yellow crust ooze       3
132             prognosis       5

[133 rows x 2 columns]

In [29]:
vals = dataset.values
vals

array([['Fungal infection', np.int64(1), np.int64(3), ..., 0, 0, 0],
       ['Fungal infection', np.int64(3), np.int64(4), ..., 0, 0, 0],
       ['Fungal infection', np.int64(1), np.int64(4), ..., 0, 0, 0],
       ...,
       ['Impetigo', np.int64(3), np.int64(7), ..., '0', '0', '0'],
       ['Impetigo', np.int64(3), np.int64(7), ..., '0', '0', '0'],
       ['Impetigo', np.int64(3), np.int64(2), ..., '0', '0', '0']],
      dtype=object)

In [30]:
#Encoding symptoms with the weights given
vals = dataset.values

#storing unique symptom names in a list called symptoms
symptoms = severity['Symptom'].unique()

for i in range(len(symptoms)):
    vals[vals == symptoms[i]] = severity[severity['Symptom'] == symptoms[i]]['weight'].values[0]

vals

array([['Fungal infection', np.int64(1), np.int64(3), ..., 0, 0, 0],
       ['Fungal infection', np.int64(3), np.int64(4), ..., 0, 0, 0],
       ['Fungal infection', np.int64(1), np.int64(4), ..., 0, 0, 0],
       ...,
       ['Impetigo', np.int64(3), np.int64(7), ..., '0', '0', '0'],
       ['Impetigo', np.int64(3), np.int64(7), ..., '0', '0', '0'],
       ['Impetigo', np.int64(3), np.int64(2), ..., '0', '0', '0']],
      dtype=object)

In [32]:
# Giving symptom weights to every symptoms from symptom severity dataset
cols = dataset.columns
dataset = pd.DataFrame(vals,columns=cols)
dataset

Disease Symptom_1 Symptom_2            Symptom_3  \
0     Fungal infection         1         3                    4   
1     Fungal infection         3         4  dischromic  patches   
2     Fungal infection         1         4  dischromic  patches   
3     Fungal infection         1         3  dischromic  patches   
4     Fungal infection         1         3                    4   
...                ...       ...       ...                  ...   
4920  Fungal infection         3         1                    4   
4921  Fungal infection         1         3                    4   
4922          Impetigo         3         7                    4   
4923          Impetigo         3         7                    4   
4924          Impetigo         3         2                    4   

                Symptom_4 Symptom_5 Symptom_6 Symptom_7 Symptom_8 Symptom_9  \
0     dischromic  patches         0         0         0         0         0   
1                       0         0         0         0         0         0   
2                       0         0         0         0         0         0   
3                       0         0         0         0         0         0   
4                       0         0         0         0         0         0   
...                   ...       ...       ...       ...       ...       ...   
4920                    0         0         0         0         0         0   
4921                    0         0         0         0         0         0   
4922                    2         3         0         0         0         0   
4923                    2         3         0         0         0         0   
4924                    7         3         0         0         0         0   

     Symptom_10 Symptom_11 Symptom_12 Symptom_13 Symptom_14 Symptom_15  \
0             0          0          0          0          0          0   
1             0          0          0          0          0          0   
2             0          0          0          0          0          0   
3             0          0          0          0          0          0   
4             0          0          0          0          0          0   
...         ...        ...        ...        ...        ...        ...   
4920          0          0          0          0          0          0   
4921          0          0          0          0          0          0   
4922          0          0          0          0          0          0   
4923          0          0          0          0          0          0   
4924          0          0          0          0          0          0   

     Symptom_16 Symptom_17  
0             0          0  
1             0          0  
2             0          0  
3             0          0  
4             0          0  
...         ...        ...  
4920          0          0  
4921          0          0  
4922          0          0  
4923          0          0  
4924          0          0  

[4925 rows x 18 columns]

In [35]:
temp = []
for i in range(1,18):
  temp.append(dataset[f'Symptom_{i}'].value_counts())

temp

[Symptom_1
 3    1930
 4    1074
 5     972
 1     679
 6     120
 2     120
 7      30
 Name: count, dtype: int64,
 Symptom_2
 3                      1699
 5                      1500
 4                      1116
 7                       350
 2                       133
 6                       108
 foul smell of urine      18
 1                         1
 Name: count, dtype: int64,
 Symptom_3
 4                      1979
 3                      1014
 5                       852
 7                       378
 2                       360
 6                       222
 foul smell of urine      84
 dischromic  patches      36
 Name: count, dtype: int64,
 Symptom_4
 4                      1110
 5                       900
 3                       846
 2                       704
 7                       553
 0                       348
 6                       348
 dischromic  patches      72
 spotting  urination      42
 0                         2
 Name: count, dtype: int64,
 Symptom_5
 4

In [41]:
dataset['Symptom_3'].unique()

array([np.int64(4), 'dischromic  patches', np.int64(3), np.int64(5),
       np.int64(6), np.int64(7), np.int64(2), 'foul smell of urine'],
      dtype=object)

In [40]:
dataset

Disease Symptom_1 Symptom_2            Symptom_3  \
0     Fungal infection         1         3                    4   
1     Fungal infection         3         4  dischromic  patches   
2     Fungal infection         1         4  dischromic  patches   
3     Fungal infection         1         3  dischromic  patches   
4     Fungal infection         1         3                    4   
...                ...       ...       ...                  ...   
4920  Fungal infection         3         1                    4   
4921  Fungal infection         1         3                    4   
4922          Impetigo         3         7                    4   
4923          Impetigo         3         7                    4   
4924          Impetigo         3         2                    4   

                Symptom_4 Symptom_5 Symptom_6 Symptom_7 Symptom_8 Symptom_9  \
0     dischromic  patches         0         0         0         0         0   
1                       0         0         0         0         0         0   
2                       0         0         0         0         0         0   
3                       0         0         0         0         0         0   
4                       0         0         0         0         0         0   
...                   ...       ...       ...       ...       ...       ...   
4920                    0         0         0         0         0         0   
4921                    0         0         0         0         0         0   
4922                    2         3         0         0         0         0   
4923                    2         3         0         0         0         0   
4924                    7         3         0         0         0         0   

     Symptom_10 Symptom_11 Symptom_12 Symptom_13 Symptom_14 Symptom_15  \
0             0          0          0          0          0          0   
1             0          0          0          0          0          0   
2             0          0          0          0          0          0   
3             0          0          0          0          0          0   
4             0          0          0          0          0          0   
...         ...        ...        ...        ...        ...        ...   
4920          0          0          0          0          0          0   
4921          0          0          0          0          0          0   
4922          0          0          0          0          0          0   
4923          0          0          0          0          0          0   
4924          0          0          0          0          0          0   

     Symptom_16 Symptom_17  
0             0          0  
1             0          0  
2             0          0  
3             0          0  
4             0          0  
...         ...        ...  
4920          0          0  
4921          0          0  
4922          0          0  
4923          0          0  
4924          0          0  

[4925 rows x 18 columns]

In [42]:
dataset = dataset.replace('spotting  urination',0)
dataset = dataset.replace('dischromic  patches',0)
dataset = dataset.replace('foul smell of urine',0)
dataset


<ipython-input-42-8b334fafce27>:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dataset = dataset.replace('spotting  urination',0)
<ipython-input-42-8b334fafce27>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dataset = dataset.replace('foul smell of urine',0)


Disease  Symptom_1  Symptom_2  Symptom_3 Symptom_4 Symptom_5  \
0     Fungal infection          1          3          4         0         0   
1     Fungal infection          3          4          0         0         0   
2     Fungal infection          1          4          0         0         0   
3     Fungal infection          1          3          0         0         0   
4     Fungal infection          1          3          4         0         0   
...                ...        ...        ...        ...       ...       ...   
4920  Fungal infection          3          1          4         0         0   
4921  Fungal infection          1          3          4         0         0   
4922          Impetigo          3          7          4         2         3   
4923          Impetigo          3          7          4         2         3   
4924          Impetigo          3          2          4         7         3   

     Symptom_6 Symptom_7 Symptom_8 Symptom_9 Symptom_10 Symptom_11 Symptom_12  \
0            0         0         0         0          0          0          0   
1            0         0         0         0          0          0          0   
2            0         0         0         0          0          0          0   
3            0         0         0         0          0          0          0   
4            0         0         0         0          0          0          0   
...        ...       ...       ...       ...        ...        ...        ...   
4920         0         0         0         0          0          0          0   
4921         0         0         0         0          0          0          0   
4922         0         0         0         0          0          0          0   
4923         0         0         0         0          0          0          0   
4924         0         0         0         0          0          0          0   

     Symptom_13 Symptom_14 Symptom_15 Symptom_16 Symptom_17  
0             0          0          0          0          0  
1             0          0          0          0          0  
2             0          0          0          0          0  
3             0          0          0          0          0  
4             0          0          0          0          0  
...         ...        ...        ...        ...        ...  
4920          0          0          0          0          0  
4921          0          0          0          0          0  
4922          0          0          0          0          0  
4923          0          0          0          0          0  
4924          0          0          0          0          0  

[4925 rows x 18 columns]

In [45]:
print("Total symptoms used to diagnose the disease : ",len(severity['Symptom'].unique()))
print("Diseases detectable by the system : ",len(dataset['Disease'].unique()))

Total symptoms used to diagnose the disease :  132
Diseases detectable by the system :  41


**Splitting datasets into train and test set**

In [46]:
dataset.tail()

Disease  Symptom_1  Symptom_2  Symptom_3 Symptom_4 Symptom_5  \
4920  Fungal infection          3          1          4         0         0   
4921  Fungal infection          1          3          4         0         0   
4922          Impetigo          3          7          4         2         3   
4923          Impetigo          3          7          4         2         3   
4924          Impetigo          3          2          4         7         3   

     Symptom_6 Symptom_7 Symptom_8 Symptom_9 Symptom_10 Symptom_11 Symptom_12  \
4920         0         0         0         0          0          0          0   
4921         0         0         0         0          0          0          0   
4922         0         0         0         0          0          0          0   
4923         0         0         0         0          0          0          0   
4924         0         0         0         0          0          0          0   

     Symptom_13 Symptom_14 Symptom_15 Symptom_16 Symptom_17  
4920          0          0          0          0          0  
4921          0          0          0          0          0  
4922          0          0          0          0          0  
4923          0          0          0          0          0  
4924          0          0          0          0          0

In [52]:
np.random.seed(42)

In [53]:
X = dataset.iloc[:,1:].values
y = dataset['Disease'].values

In [54]:
X

array([[1, 3, 4, ..., 0, 0, 0],
       [3, 4, 0, ..., 0, 0, 0],
       [1, 4, 0, ..., 0, 0, 0],
       ...,
       [3, 7, 4, ..., '0', '0', '0'],
       [3, 7, 4, ..., '0', '0', '0'],
       [3, 2, 4, ..., '0', '0', '0']], dtype=object)

In [55]:
y

array(['Fungal infection', 'Fungal infection', 'Fungal infection', ...,
       'Impetigo', 'Impetigo', 'Impetigo'], dtype=object)

In [56]:
X_train,X_test,y_train,y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

# Model : Using Random Forest Classifier

In [60]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

In [62]:
model = RandomForestClassifier()

#Train the model
model.fit(X_train,y_train)

#Prediction
y_predict = model.predict(X_test)

#Accuracy
print("Accuracy : ",(metrics.accuracy_score(y_test,y_predict))*100)

Accuracy :  99.59390862944163


In [63]:
print(classification_report(y_test,y_predict))

                                         precision    recall  f1-score   support

(vertigo) Paroymsal  Positional Vertigo       1.00      1.00      1.00        27
                                   AIDS       1.00      1.00      1.00        24
                                   Acne       1.00      1.00      1.00        21
                    Alcoholic hepatitis       1.00      1.00      1.00        24
                                Allergy       0.92      1.00      0.96        24
                              Arthritis       1.00      1.00      1.00        25
                       Bronchial Asthma       1.00      1.00      1.00        27
                   Cervical spondylosis       1.00      0.90      0.95        21
                            Chicken pox       1.00      1.00      1.00        20
                    Chronic cholestasis       1.00      1.00      1.00        25
                            Common Cold       1.00      1.00      1.00        19
                           

# With an accuracy of 99.59%, our model delivers highly reliable predictions.